add git sparse checkout

mobileoverlord opened the pull request:
This is a start to adding sparse checkout to the GIT SCM.

Sparse checkout can be used for picking a dep from a repo which may contain many deps in sub folders

For example.

The git repository
https://github.com/nerves-project/toolchains
contains many toolchains including 
`nerves_toolchain_armv5tejl_unknown_linux_musleabi`

To use git and single out that subfolder int he repo as the dependency destination we could use sparse

`{:nerves_toolchain_armv5tejl_unknown_linux_musleabi, github: "nerves-project/toolchains", sparse: "nerves_toolchain_armv5tejl_unknown_linux_musleabi"}`

mobileoverlord commented:
Things to do yet

- [ ] verify update / locks
- [ ] Verify git version minimum for sparse checkout (1.7.0)
- [ ] Add tests

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -25,17 +25,24 @@ defmodule Mix.SCM.Git do
+        true ->
+          nil
+      end
+    if opts[:sparse] do
```
You can store the `:dest` option as `:checkout` or `:git_dest` so you don't need to use replace_suffix later on.

josevalim commented:
One of the scenarios we will need to test is using the same checkout and moving from sparse to non-sparse and vice-versa (i.e. add and remove the sparse option, does it still work?)

ericmj commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -111,15 +132,27 @@ defmodule Mix.SCM.Git do
       [_] -> opts
       _   ->
-        Mix.raise "You should specify only one of branch, ref or tag, and only once. " <>
+        Mix.raise "You should specify only one of sparse, branch, ref or tag, and only once. " <>
```
You cannot use sparse with the other options?

mobileoverlord commented:
I agree @ericmj I misunderstood the intention. I have changed it to revert that test and I have also added the ability to define multiple sparse options as the sparse function allows this. I haven't figured out a reason you would need to do this to satisfy a dependency, but, I also don't know of a reason not to allow it.

As per testing turning it off and on and updating. I didn't see any active git tests in the suite. Should I just try these scenarios out, or do we want to use this time to make additional tests?

mobileoverlord commented:
hum thats curious, those tests didn't fail for me

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -120,6 +154,21 @@ defmodule Mix.SCM.Git do
+      File.cwd!
+      |> Path.join(".git/info/sparse-checkout")
+
+    Keyword.get_values(opts, :sparse)
```
I don't think we should allow multiple sparse because `dest` will point to only one of them.

josevalim commented on lib/mix/lib/mix/scm/git.ex:
```diff
@@ -120,6 +154,21 @@ defmodule Mix.SCM.Git do
+  defp init_sparse(opts) do
+    git!("init --quiet")
+    git!("remote add origin #{opts[:git]} --fetch")
+    git!("config core.sparsecheckout true")
```
We need to set this to false when we do a update and it is not sparse.

mobileoverlord commented:
I don't understand why I am getting different test results. I don;t get this failure when I run `make test` but I do get other failures 

```
test uniq/1 (EnumTest.Range)
     test/elixir/enum_test.exs:1236
     ** (UndefinedFunctionError) function Stream.Reducers.uniq_by/1 is undefined or private. Did you mean one of:

           * uniq_by/1
           * uniq_by/2

     stacktrace:
       (elixir) Stream.Reducers.uniq_by(#Function<187.4743325/1 in Enum.uniq/1>)
       (elixir) lib/enum.ex:2489: Enum.uniq_by/2
       test/elixir/enum_test.exs:1237: (test)
```

mobileoverlord commented:
I figured out the tests :)

mobileoverlord commented:
so I added a helper which chooses :git_dest if it exists and falls back to :dest. I talked with @ericmj and I am not sure which approach is better, this or always having :git_dest. We would have to fix a handful of tests to include the git_dest key.

mobileoverlord commented:
Also, I've been using this sample for testing
https://github.com/mobileoverlord/sparse
